Index: scripts/Download
===================================================================
--- scripts/Download	(revision 72856)
+++ scripts/Download	(working copy)
@@ -435,7 +435,7 @@
 
 		    if [ -n "$mirror" -a "$mirror" != "none" -a -z "${gzfile##download/mirror/*}" ]; then
 			# try to use mirror
-			if ! download_file_now "!$mirror/${gzfile#download/mirror/}" $gzfile; then
+			if ! download_file_now "!$mirror/${gzfile#download/mirror/}" $gzfile "$location"; then
 				echo "INFO: Download from mirror failed, trying original URL."
 				download_file_now "$location" $gzfile || downloaderror=1
 			fi
@@ -484,13 +484,15 @@
 # This function executes the actual download using curl.
 #
 download_file_now() {
-	local location="$1" gzfile="$2" curlret=0
+	local location="$1" gzfile="$2" orig_location="$3"
+	local curl_header= curlret=0
 
 	# Create URL
 	#
 	case "$location" in
 		manual://*) url="$location"	;;
-		!*)	url="${location#!}"	;;
+		!*)	curl_header="X-Orig-URL: ${orig_location%/*}/${gzfile##*/}"
+			url="${location#!}"	;;
 		*)	url="${location%/*}/${gzfile##*/}"	;;
 	esac
 
@@ -536,8 +538,7 @@
 			url="$trurl"
 		fi
 		unset trurl trfile
-
-		curl --progress-bar $resume $curl_options "$url" -o "$gzfile.incomplete"
+		curl --progress-bar $resume $curl_options -H "$curl_header" "$url" -o "$gzfile.incomplete"
 		curlret="$?"
 
 		if [ "$resume" ] &&
